tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/DefineConstantsPropertyPageProviderTests.cs

```diff
@@ -0,0 +1,195 @@
+                }
+            }
+            if (valid.Length > 0)
+            {
+                return valid.Remove(valid.Length - 1);                          //remove the trailing comma
+            }
+            else
+            {
+                return valid;
+            }
+        }
+
+        //determines if input is in valid format
+        private bool ValidInput(string input)
+        {
+            var regex = new Regex(@"(\S+?\s*)\=(\S+?\s*)");
+            return regex.IsMatch(input);
+        }
+
+        //determines if encoding is in valid format
+        private bool ValidEncoding(string encoding)
+        {
+            var regex = new Regex(@"(\S+?\s*)\=[\x22](\S+?\s*)[\x22],?");
+            return regex.IsMatch(encoding);
+        }
+
+        // outputFormat true for the format key="value", false for key=value
+        private string EncodeOutputFormat(string input, bool outputFormat)
+        {
+            input += ",";
+            var regex = new Regex(@"(\S+?\s*)\=(\S+?\s*),");
+            string encoded = "";
+            foreach (System.Text.RegularExpressions.Match match in regex.Matches(input))
+            {
+                string key = match.Groups[1].Value.Replace("/", "//").Replace("=", "/=").Replace(",", "/,");
+                string value = match.Groups[2].Value.Replace("/", "//").Replace("=", "/=").Replace(",", "/,");
+                encoded += outputFormat ? key + "=\"" + value + "\"," : key + "=" + value + ",";
+            }
+            return encoded.TrimEnd(',');
+        }
+
+        //parses repeated keys inputted key value pairs
+        private string getDefinedRepeat(string input)
+        {
+            input += ",";
+            List<KeyValuePair<string, string>> uniqueKeyValuePairs = new List<KeyValuePair<string, string>>();
+            var regex = new Regex(@"(\S+?\s*)\=(\S+?\s*),");
+            foreach (System.Text.RegularExpressions.Match match in regex.Matches(input))
+            {
+                string key = match.Groups[1].Value;
+                string value = match.Groups[2].Value;
+                var index = uniqueKeyValuePairs.FindIndex(kvp => kvp.Key == key);
+                if (index >= 0)
+                {
+                    uniqueKeyValuePairs[index] = new KeyValuePair<string, string>(key, value);
+                }
+                else
+                {
+                    uniqueKeyValuePairs.Add(new KeyValuePair<string, string>(key, value));
+                }
+            }
+
+            var formattedInput = "";
+            foreach (var kvp in uniqueKeyValuePairs)
+            {
+                formattedInput += kvp.Key + "=" + kvp.Value + ",";
+            }
+
+            if (formattedInput.Length > 0)
+            {
+                return formattedInput.Remove(formattedInput.Length - 1);           //remove the trailing comma
+            }
+            else
+            {
+                return formattedInput;
+            }
```

Comment:
This would be more efficient if it used a `StringBuilder`.

In the `foreach` loop, each iteration creates a new temporary string of increasing size. Consider pairs of `[A,1],[B,2],[C,3]`. In the current code we'd allocate all these strings:

- `"A=1,"`
- `"B=2,"`
- `"A=1,B=2,"`
- `"C=3,"`
- `"A=1,B=2,C=3,"`
- `"A=1,B=2,C=3"`

It is possible to allocate only the last one using `StringBuilder`. `StringBuilder` internally creates a buffer to write data into which can avoid a lot of allocations.

Something like:

```suggestion
            StringBuilder formattedInput = new();
            foreach (var kvp in uniqueKeyValuePairs)
            {
                formattedInput.Append(kvp.Key).Append('=').Append(kvp.Value).Append(',');
            }

            if (formattedInput.Length > 0)
            {
                return formattedInput.ToString(0, formattedInput.Length - 1);           //remove the trailing comma
            }
            else
            {
                return formattedInput.ToString();
            }
```